pkg/util/pki/asn1_util_test.go

```diff
@@ -0,0 +1,240 @@
+		if err != nil {
+			if tc.expectedErr == nil {
+				t.Errorf("Unexpected error: %v", err)
+			} else if err.Error() != tc.expectedErr.Error() {
+				t.Errorf("Expected error: %v, got: %v", tc.expectedErr, err)
+			}
+		} else if !oid.Equal(tc.expectedOid) {
+			t.Errorf("Expected OID: %v, got: %v", tc.expectedOid, oid)
+		}
+	}
+}
+
+func TestMarshalAndUnmarshalUniversalValue(t *testing.T) {
+	testCases := []struct {
+		name                string
+		uv                  UniversalValue
+		raw                 asn1.RawValue
+		overrideRoundtripUv *UniversalValue
+	}{
+		{
+			name: "Test with IA5String",
+			uv: UniversalValue{
+				Ia5String: "test",
+			},
+			raw: asn1.RawValue{
+				Bytes: []byte("test"),
+				Class: asn1.ClassUniversal,
+				Tag:   asn1.TagIA5String,
+			},
+		},
+		{
+			name: "Test with Utf8String",
+			uv: UniversalValue{
+				Utf8String: "test",
+			},
+			raw: asn1.RawValue{
+				Bytes: []byte("test"),
+				Class: asn1.ClassUniversal,
+				Tag:   asn1.TagUTF8String,
+			},
+		},
+		{
+			name: "Test with PrintableString",
+			uv: UniversalValue{
+				PrintableString: "test",
+			},
+			raw: asn1.RawValue{
+				Bytes: []byte("test"),
+				Class: asn1.ClassUniversal,
+				Tag:   asn1.TagPrintableString,
+			},
+		},
+		{
+			name: "Test with Bytes",
+			uv: UniversalValue{
+				Bytes: []byte{0x16, 0x04, 0x74, 0x65, 0x73, 0x74},
+			},
+			overrideRoundtripUv: &UniversalValue{
+				Ia5String: "test",
+			},
+			raw: asn1.RawValue{
+				Bytes: []byte("test"),
+				Class: asn1.ClassUniversal,
+				Tag:   asn1.TagIA5String,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			{
+				rawValue, err := MarshalUniversalValue(tc.uv)
+				if err != nil {
+					t.Errorf("Unexpected error: %v", err)
+				}
+
+				// Calculate fullBytes
+				fullBytes, err := asn1.Marshal(tc.raw)
+				if err != nil {
+					t.Errorf("Unexpected error: %v", err)
+				}
+
+				if !reflect.DeepEqual(rawValue, fullBytes) {
+					t.Errorf("Expected rawValue: %v, got: %v", fullBytes, rawValue)
+				}
+			}
+
+			{
+				uv, err := UnmarshalUniversalValue(tc.raw)
+				if err != nil {
+					t.Errorf("Unexpected error: %v", err)
+				}
+
+				targetUv := tc.uv
+				if tc.overrideRoundtripUv != nil {
+					targetUv = *tc.overrideRoundtripUv
+				}
+				if !reflect.DeepEqual(uv, targetUv) {
+					t.Errorf("Expected uv: %v, got: %v", targetUv, uv)
+				}
+			}
+		})
+	}
+}
+
+func TestIsIA5String(t *testing.T) {
+	ia5Strings := []string{
+		"test",
+		"1234",
+		"!@#$",
+		" ",
+		"",
+	}
+
+	for _, ia5String := range ia5Strings {
+		err := isIA5String(ia5String)
+
+		if err != nil {
+			t.Errorf("Expected IA5 string %q, got: %w", ia5String, err)
+		}
+	}
+
+	nonIA5Strings := []string{
+		"中文",
+	}
+
+	for _, nonIA5String := range nonIA5Strings {
+		err := isIA5String(nonIA5String)
+
+		if err != nil {
+			t.Errorf("Expected non-IA5 string %q, got: %w", nonIA5String, err)
+		}
+	}
+}
+
+func TestIsPrintable(t *testing.T) {
```

Comment:
suggestion: I think we could benefit from testing UTF-8 strings too. There are [lists of invalid utf-8 string](https://stackoverflow.com/questions/1301402/example-invalid-utf8-string) out there which we could use.

Reply:
also done